File: appcore/src/main/java/mobi/cangol/mobile/http/polling/PollingResponseHandler.java
/**
 * Copyright (c) 2013 Cangol
 * <p>
 * Licensed under the Apache License, Version 2.0 (the "License")
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package mobi.cangol.mobile.http.polling;

import android.os.Handler;
import android.os.Looper;
import android.os.Message;

import java.io.IOException;

import okhttp3.Response;
import okhttp3.ResponseBody;

public class PollingResponseHandler {
    protected static final int START_MESSAGE = -1;
    protected static final int SUCCESS_MESSAGE = 0;
    protected static final int FAILURE_MESSAGE = 1;
    protected static final int FINISH_MESSAGE = 2;
    private final static String TAG = PollingHttpClient.TAG;
    private final static boolean DEBUG = true;
    private Handler handler;

    public PollingResponseHandler() {
        if (Looper.myLooper() != null) {
            handler = new Handler() {
                public void handleMessage(Message msg) {
                    PollingResponseHandler.this.handleMessage(msg);
                }
            };
        }
    }

    public boolean isFailResponse(String content) {
        return false;
    }

    public void onStart() {
    }

    public void onPollingFinish(int execTimes, String content) {
    }

    public void onSuccess(int statusCode, String content) {
    }

    public void onFailure(Throwable error, String content) {
    }

    public void sendStartMessage() {
        sendMessage(obtainMessage(START_MESSAGE, new Object[]{new Integer(-1), "exec start"}));
    }

    public void sendFinishMessage(int execTimes) {
        sendMessage(obtainMessage(FINISH_MESSAGE, new Object[]{new Integer(execTimes), "exec finish"}));
    }

    public void sendSuccessMessage(int statusCode, String responseBody) {
        sendMessage(obtainMessage(SUCCESS_MESSAGE, new Object[]{new Integer(statusCode), responseBody}));
    }

    public void sendFailureMessage(IOException e, String responseBody) {
        sendMessage(obtainMessage(FAILURE_MESSAGE, new Object[]{e, responseBody}));
    }

    boolean sendResponseMessage(Response response) {
        boolean result = false;
        ResponseBody responseBody=response.body();
        String content=null;
        if(response.isSuccessful()){
            if (responseBody != null) {
                content = responseBody.toString();
            }
            if (isFailResponse(content)) {
                sendFailureMessage(new IOException("code="+response.code()), content);
                result = false;
            } else {
                sendSuccessMessage(response.code(), content);
                result = true;
            }
        }else{
            sendFailureMessage(new IOException("code="+response.code()), content);
            result = false;
        }

        return result;
    }

    protected void handleMessage(Message msg) {
        Object[] response;
        switch (msg.what) {
            case SUCCESS_MESSAGE:
                response = (Object[]) msg.obj;
                handleSuccessMessage(((Integer) response[0]).intValue(), (String) response[1]);
                break;
            case FAILURE_MESSAGE:
                response = (Object[]) msg.obj;
                handleFailureMessage((Throwable) response[0], (String) response[1]);
                break;
            case FINISH_MESSAGE:
                response = (Object[]) msg.obj;
                handleFinishMessage(((Integer) response[0]).intValue(), (String) response[1]);
                break;
            case START_MESSAGE:
                handleStartMessage();
                break;
        }
    }

    protected void handleStartMessage() {
        onStart();
    }

    protected void handleFinishMessage(int execTimes, String responseBody) {
        onPollingFinish(execTimes, responseBody);
    }

    protected void handleSuccessMessage(int statusCode, String responseBody) {
        onSuccess(statusCode, responseBody);
    }

    protected void handleFailureMessage(Throwable e, String responseBody) {
        onFailure(e, responseBody);
    }

    protected void sendMessage(Message msg) {
        if (handler != null) {
            handler.sendMessage(msg);
        } else {
            handleMessage(msg);
        }
    }

    protected Message obtainMessage(int responseMessage, Object response) {
        Message msg = null;
        if (handler != null) {
            msg = this.handler.obtainMessage(responseMessage, response);
        } else {
            msg = new Message();
            msg.what = responseMessage;
            msg.obj = response;
        }
        return msg;
    }
}

File: appcore/src/main/java/mobi/cangol/mobile/http/polling/PollingResponseHandler.kt
/**
 * Copyright (c) 2013 Cangol
 *
 *
 * Licensed under the Apache License, Version 2.0 (the "License")
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package mobi.cangol.mobile.http.polling

import android.os.Handler
import android.os.Looper
import android.os.Message
import mobi.cangol.mobile.logging.Log
import okhttp3.Response
import java.io.IOException

open class PollingResponseHandler {
    private var handler: Handler? = null

    init {
        if (Looper.myLooper() != null) {
            handler = object : Handler() {
                override fun handleMessage(msg: Message) {
                    this@PollingResponseHandler.handleMessage(msg)
                }
            }
        }
    }

    /**
     * 是否结束请求
     *
     * @param content
     * @return
     */
    open fun isFailResponse(content: String?): Boolean {
        return false
    }

    /**
     * 启动开始
     */
    open fun onStart() {
        //do nothings
    }

    /**
     * 轮询结束
     *
     * @param execTimes
     * @param content
     */
    open fun onPollingFinish(execTimes: Int, content: String) {
        //do nothings
    }

    /**
     * 轮询成功
     *
     * @param statusCode
     * @param content
     */
    open fun onSuccess(statusCode: Int, content: String) {
        //do nothings
    }

    /**
     * 轮询失败
     *
     * @param error
     * @param content
     */
    open fun onFailure(error: Throwable, content: String) {
        //do nothings
    }

    fun sendStartMessage() {
        sendMessage(obtainMessage(START_MESSAGE, arrayOf(-1, "exec start")))
    }

    fun sendFinishMessage(execTimes: Int) {
        sendMessage(obtainMessage(FINISH_MESSAGE, arrayOf(execTimes, "exec finish")))
    }

    protected fun sendSuccessMessage(statusCode: Int, responseBody: String?) {
        sendMessage(responseBody?.let { arrayOf(statusCode, it) }?.let { obtainMessage(SUCCESS_MESSAGE, it) })
    }

    fun sendFailureMessage(e: IOException, responseBody: String?) {
        sendMessage(responseBody?.let { arrayOf<Any>(e, it) }?.let { obtainMessage(FAILURE_MESSAGE, it) })
    }

    fun sendResponseMessage(response: Response): Boolean {
        var result:Boolean
        val responseBody = response.body()
        var content: String? = null
        if (response.isSuccessful) {
            if (responseBody != null) {
                try {
                    content = responseBody.string()
                } catch (e: IOException) {
                    Log.e(e.message)
                }

            }
            if (isFailResponse(content)) {
                sendFailureMessage(IOException("code=" + response.code()), content)
                result = false
            } else {
                sendSuccessMessage(response.code(), content)
                result = true
            }
        } else {
            sendFailureMessage(IOException("code=" + response.code()), content)
            result = false
        }

        return result
    }

    protected fun handleMessage(msg: Message) {
        val response: Array<Any>
        when (msg.what) {
            SUCCESS_MESSAGE -> {
                response = msg.obj as Array<Any>
                handleSuccessMessage((response[0] as Int).toInt(), response[1] as String)
            }
            FAILURE_MESSAGE -> {
                response = msg.obj as Array<Any>
                handleFailureMessage(response[0] as Throwable, response[1] as String)
            }
            FINISH_MESSAGE -> {
                response = msg.obj as Array<Any>
                handleFinishMessage((response[0] as Int).toInt(), response[1] as String)
            }
            START_MESSAGE -> handleStartMessage()
            else -> {
            }
        }
    }

    protected fun handleStartMessage() {
        onStart()
    }

    protected fun handleFinishMessage(execTimes: Int, responseBody: String) {
        onPollingFinish(execTimes, responseBody)
    }

    protected fun handleSuccessMessage(statusCode: Int, responseBody: String) {
        onSuccess(statusCode, responseBody)
    }

    protected fun handleFailureMessage(e: Throwable, responseBody: String) {
        onFailure(e, responseBody)
    }

    protected fun sendMessage(msg: Message?) {
        if (handler != null) {
            handler!!.sendMessage(msg)
        } else {
            handleMessage(msg!!)
        }
    }

    protected fun obtainMessage(responseMessage: Int, response: Any): Message? {
        var msg: Message?
        if (handler != null) {
            msg = this.handler!!.obtainMessage(responseMessage, response)
        } else {
            msg = Message()
            msg.what = responseMessage
            msg.obj = response
        }
        return msg
    }

    companion object {
        protected const val START_MESSAGE = -1
        protected const val SUCCESS_MESSAGE = 0
        protected const val FAILURE_MESSAGE = 1
        protected const val FINISH_MESSAGE = 2
    }
}